Websocket connection is active.

jesseobrien opened the pull request:
- Connects on the front end and requests the account info
- Requires the self-signed cert to be accepted in the browser for the
socket to connect
- Sends back account info to initialize the page

jesseobrien committed:
Websocket connection is active.

- Connects on the front end and requests the account info
- Requires the self-signed cert to be accepted in the browser for the
socket to connect
- Sends back account info to initialize the page

jesseobrien committed:
Websocket updates and account upgrades are done.

- Front end is updating when new account updates go through.
- Account upgrade button is upgrading the account properly.

jesseobrien committed:
Adding dockerfile and run commands

jesseobrien committed:
Properly locking UpdateStore and fixing test

- Fixing test dependencies
- Ensuring the proper mutex is being locked by subscribers when they
unsubscribe and clean up the lists of subscribers

jesseobrien committed:
Moving to rollup js

jesseobrien committed:
Putting in a javascript test for the login function

awly commented on Makefile:
```diff
@@ -1,6 +1,7 @@
 .ONESHELL:
 .PHONY: install
 install:
+	$(MAKE) certs
```
```suggestion
install: certs
```

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+
+func AccountUpdatesHandler(tokenStore *storage.TokenStore, accountStore storage.AccountStore, accountUpdateStore *storage.AccountUpdateStore) echo.HandlerFunc {
+	return func(ctx echo.Context) error {
+		updateRequest := AccountUpdatesRequest{}
```
do you need a struct for this parameter binding, or could you use `var accountID string` instead here?

jesseobrien replied:
Yeah it doesn't require a struct. Would save on the allocation.

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+		return fmt.Errorf("could not get account info %w", err)
+	}
+
+	jsonInfo, _ := json.Marshal(accountInfo)
```
handle the error

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+	for {
+		accountUpdate := <-subscription.Updates
+
+		jsonUpdate, _ := json.Marshal(accountUpdate)
```
handle the error

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+	defer subscription.Unsubscribe()
+
+	for {
+		accountUpdate := <-subscription.Updates
```
```suggestion
	for accountUpdate := range subscription.Updates {	
```

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+		jsonUpdate, _ := json.Marshal(accountUpdate)
+
+		if err := socket.WriteJSON(SocketMessage{Operation: AccountMetricsUpdated, Data: string(jsonUpdate)}); err != nil {
+			ctx.Logger().Error(err)
```
```suggestion
			ctx.Logger().Error(err)
			return
```

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func checkMessageToken(ctx echo.Context, msg SocketMessage, tokenStore *storage.TokenStore, socket *websocket.Conn) error {
+	token := strings.TrimPrefix(msg.Token, "Bearer ")
```
don't add the `Bearer` prefix to the token, since this is not the `Authorization` header

awly commented on internal/http/login.go:
```diff
@@ -8,13 +8,16 @@ import (
 )
 
+const UserAccountId = ""
+
```
```suggestion
```

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,138 @@
+			if err := checkMessageToken(ctx, msg, tokenStore, socket); err != nil {
+				ctx.Logger().Error(err)
+				break
+			}
```
check that the token matches `updateRequest.AccountId`

awly commented on internal/storage/account_updates.go:
```diff
@@ -0,0 +1,75 @@
+	subscriptions := aus.accountUpdates[accountId]
+
+	for _, subscription := range subscriptions {
+		subscription.Updates <- update
```
if the receiving end is blocked, the entire Fanout will be stuck

awly commented on internal/storage/account_updates.go:
```diff
@@ -0,0 +1,75 @@
+	subscriptionList *[]AccountUpdateSubscription
+
+	// Link to the top level mutex in the update store to lock when we're unsubscribing
+	mu *sync.Mutex
```
spreading subscription state management over multiple types by sharing memory sounds sketchy.

ideally, make `Unsubscribe` a method of `AccountUpdateStore`:
```
s := store.Subscribe(accountID)
defer store.Unsubscribe(s)
```

jesseobrien replied:
I was initially going to go that path. I decided that it would make more sense to go the more idiomatic route of something like connections, etc where you get the object back and have it close() itself. I agree with you however, sharing the memory is a strange pattern. I'll swtich it up.

awly commented on internal/storage/account_updates.go:
```diff
@@ -0,0 +1,75 @@
+
+func (aus *AccountUpdateStore) Fanout(accountId string, update AccountUpdate) {
+	// Get the subscriptions, if there are none, the range will no-op
+	subscriptions := aus.accountUpdates[accountId]
```
lock the store mutex

awly commented on internal/storage/accounts.go:
```diff
@@ -0,0 +1,26 @@
+	Id         string          `json:"id"`
+	PlanType   AccountPlanType `json:"plan_type"`
+	LoginCount int             `json:"login_count"`
+	PlanLimit  int             `json:"plan_limit"`
```
```suggestion
	PlanLimit  AccountPlanLimit             `json:"plan_limit"`
```

awly commented on internal/storage/accounts_pg.go:
```diff
@@ -0,0 +1,56 @@
+		info.PlanLimit = int(AccountPlanStandardLimit)
+	case AccountPlanTypeEnterprise:
+		info.PlanLimit = int(AccountPlanEnterpriseLimit)
+	}
```
return an error if none of the switch cases match

alex-kovoy commented on web/src/dashboard.js:
```diff
@@ -0,0 +1,196 @@
+const Dashboard = ({ handleLogout }) => {
+  const [account, setAccount] = useState(new AccountInfo());
+  const accountRef = useRef();
+  accountRef.current = account;
```
Do we really need accountRef? By looking at how it's been used it seems it is unnecessary as we always tend to call `setAccount` after that.

alex-kovoy commented on web/src/dashboard.js:
```diff
@@ -0,0 +1,196 @@
+          setAccountUpgraded(false);
+        }, 4000);
+      }
+    } catch (err) {
```
No error messages? :-(

alex-kovoy commented on web/src/dashboard.js:
```diff
@@ -0,0 +1,196 @@
+    () => {
+      ws.current.close();
+    };
+  }, []);
```
It would be great if we could separate rendering logic from business/service logic. You could move all data manipulation logic and service interaction into the hooks. For example something like this:

```ts

function useAccount(){
  const [ account, setAccount ] = React.useState(getDefaultState);  
  const { id, plan_limit, login_count, plan_type } = account;
  const usagePercentage = (login_count / plan_limit) * 100  
  
  const updateAccount = json => {
    setState({
      ...state,
      ...json      
    })
  }
    
  const {
    id,
    plan_limit,
    login_count,
    usagePercentage,
    cost,
    updateAccount
  } 
}

function useSession(){
  const accountId = localStorage.getItem("user_account_id");
  const token = localStorage.getItem("user_token");
  const addr = window.location;

  return {
    accountId,
    token,
    addr
  }
}

```

so then the Dashboard component can just initialize these hooks and render simple fields.

```tsx
const [ error, setError ] = useState(null);
const session = useSession();
const account = useAccount();

handleError(err){
  setError(err)
}

handleAccountUpdate(json){
  account.update(json);  
}

backend = useBackend(session, {  
  onUpdate: handleUpdate
  onError: handleError
});

handleUpgradePlanClick(){
  backend.upgrade();
}

return (
  <Dashboard>
    //
  </Dashboard>

)
```

You can also move all the hooks into a container component and inject them to inner Dashboard. This will make the Dashboard super dummy making it easier to unit-tests

```tsx
export default function Container(){
  const [ error, setError ] = useState(null);
  const session = useSession();
  const account = useAccount();

  handleError(err){
    setError()
  }

  handleAccountUpdate(json){
    account.update(json);  
  }

  backend = useBackend(session, {  
    onUpdate: handleUpdate
    onError: handleError
  });

  handleUpgradePlanClick(){
    backend.upgrade();
  }

  return (
    <Dashboard  
      { ...account }  
      onPlanUpdateClick={handleUpgradePlanClick} 
      onLogoutClick={handleLogout} ...
    />    
  )
}

```

jesseobrien replied:
I agree. I initially didn't want to make these changes so we didn't have to carry the project out too much longer but I'll make these refactors.

alex-kovoy replied:
@JesseObrien if you are not familiar with react hooks (which is fine) lets just move all hooks into 1 place `useDashboard()` and decouple it from the Dashboard component. That hook will handle all datafetching and init logic (almost what you have now).

```ts
function export default Dashboard() {   
   const { account, updatePlan, error, and_other_stuff } = useDashboard();   
   return (
     <div>
     </div>
}

jesseobrien replied:
@alex-kovoy no problem, I've already refactored the components to use custom hooks using similar patterns to what you suggested. :)

alex-kovoy commented on web/src/dashboard.js:
```diff
@@ -0,0 +1,196 @@
+          </div>
+
+          <h3>
+            Users: {account.login_count}/{account.plan_limit}
```
I would pick 1 place to define this logic. Currently some of them are in the class methods and some are inlined in jsx.

jesseobrien replied:
Agreed, consistency is :key:

jesseobrien committed:
Refactoring javascript functionality and applying fixes.

- Refactoring the subscription functionality for the account updates
- Moving all of the hook functionality into their own files under svc/
- Refactoring all of the javascript files to use the custom hooks and
sessionProvider

alex-kovoy commented on web/src/svc/use_account.js:
```diff
@@ -0,0 +1,55 @@
+};
+
+const useAccount = () => {
+  const [account, setAccount] = useState(DefaultAccountState());
```
useState can accept a function useState(fn) - this way it will execute it once and ignore it for all consequent calls. 

```js
function getDefault(){
   return {
     ...
   }
}

const [ ... ] = useState(getDefault);
```

alex-kovoy commented on web/src/svc/use_session.js:
```diff
@@ -0,0 +1,80 @@
+
+export const useSession = () => useContext(sessionContext);
+
+export const SessionProvider = ({ children }) => {
```
It does not seem that you need a provider (although provider makes it easier to test).

If you remove the context altogether and use `useProviderSession` as is, it will work without differences (based on a quick scan).

alex-kovoy reviewed (approved):
I see some things that could be improved (as always in this world) but overall I think this is fine for the challenge.

awly commented on internal/http/account_updates.go:
```diff
@@ -0,0 +1,145 @@
+		return nil
+	}
+
+	// Let the front end know there's an authorization problem
+	if err := socket.WriteJSON(SocketMessage{Operation: AuthorizationFailure, Data: "invalid auth token"}); err != nil {
+		return err
+	}
```
the caller already writes a message on error, prefer to keep all the socket handling there

awly commented on internal/storage/tokens.go:
```diff
@@ -26,6 +33,20 @@ func (ts *TokenStore) IsValid(token string) bool {
 
+func (ts *TokenStore) IsValidAccountToken(token string, accountId string) bool {
+
+	if !ts.IsValid(token) {
+		return false
+	}
+
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
```
nit: lock the mutex for the entire duration of this method
otherwise you're doing the map lookup twice and potentially looking at two distinct states of the store.
